### PR TITLE
[numpy] fix mix numpy scalar and MXNet numpy ndarray

### PR DESCRIPTION
**Pull request by BenjaminCHEN2016**
## Description ##
Attempt to resolve issue mentioned in #16743
It is done by casting the inputs into mx_np or np.
Current conversion rules:
| Expression | a type | b type | out type|
| --- | --- | --- | --- |
| `a += b` | onp | mx_np | onp |
| `a += b` | mx_np | onp | mx_np |
| `c = a + b` | onp | mx_np | onp |
| `c = a + b` | mx_np | onp | mx_np |

However, the example code in the issue still can't run due to type issue.

```python
import mxnet as mx
import numpy as np
mx.npx.set_np()
a = np.array(1) // the type in np is int64
a += mx.np.array(1) // the type in mx_np is float32
```

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at https://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by hgt312** [comment]
@yzhliu @sxjscience Any thoughts on the conversion rules?

**Comment by sxjscience** [comment]
```python
import jax.numpy as jnp
import numpy as onp
# jnp += onp
a = jnp.ones((10,))
b = onp.ones((10,))
a += b
print(type(a))

# onp += jnp
a = onp.ones((10,))
b = jnp.ones((10,))
a += b
print(type(a))
```

Output: 
```
<class 'jax.interpreters.xla.DeviceArray'>
<class 'jax.interpreters.xla.DeviceArray'>
```

**Comment by sxjscience** [comment]
In Jax, if any ndarray is a Jax ndarray, the result will also be a Jax ndarray.

**Comment by hgt312** [comment]
Jax does not implement inplace op overload like `__iadd__`. It is not a real inplace operation.
In my opinion, inplace operation should keep its type, while the other should return a mx ndarray.

**Comment by sxjscience** [comment]
> Jax does not implement inplace op overload like `__iadd__`. It is not a real inplace operation.
> In my opinion, inplace operation should keep its type, while the other should return a mx ndarray.

I agree with that.

**Comment by leezu** [comment]
We can also consider not exposing the "real" inplace operation

**Comment by BenjaminCHEN2016** [comment]
Conversion rules:
| Expression | a type | b type | out type|
| --- | --- | --- | --- |
| `a += b` | onp | mx_np | onp |
| `a += b` | mx_np | onp | mx_np |
| `c = a + b` | onp | mx_np | mx_np |
| `c = a + b` | mx_np | onp | mx_np |

**Comment by BenjaminCHEN2016** [comment]
@mxnet-bot ci [miscellaneous, unix-cpu],

**Comment by BenjaminCHEN2016** [comment]
@mxnet-bot ci [miscellaneous, unix-cpu]

**Comment by BenjaminCHEN2016** [comment]
@mxnet-bot run ci [miscellaneous, unix-cpu]

**Comment by BenjaminCHEN2016** [comment]
@mxnet-bot run ci [windows-cpu, windows-gpu, centos-cpu]

**Comment by hgt312** [comment]
LGTM now. @yzhliu

**Comment by BenjaminCHEN2016** [comment]
@mxnet-bot run ci [centos-cpu]

**Comment by BenjaminCHEN2016** [comment]
@mxnet-bot run ci [unix-gpu]

**Comment by BenjaminCHEN2016** [comment]
@mxnet-bot run ci [unix-gpu]

**Comment by BenjaminCHEN2016** [comment]
@sxjscience Any thoughts on this pr?

File: python/mxnet/numpy/multiarray.py [hunk 1]
```diff
@@ -282,6 +313,16 @@ def __array_ufunc__(self, ufunc, method, *inputs, **kwargs):  # pylint: disable=
...
                 out = onp_op(*new_inputs, **kwargs)
                 return _as_mx_np_array(out, ctx=inputs[0].ctx)
+            # ops with np mx_np
+            elif name in ufunc_list and isinstance(inputs[0], _np.ndarray):
+                # inplace
+                if 'out' in kwargs:
+                    new_inputs = [arg.asnumpy() if isinstance(arg, ndarray) else arg for arg in inputs]
+                    return onp_op(*new_inputs, **kwargs)
+                else:
+                    new_inputs = [_as_mx_np_array(arg, ctx=inputs[1].ctx)
```
Comment (sxjscience) [inline python/mxnet/numpy/multiarray.py:323]: I'm not sure why should we use `inputs[1].ctx` instead of `inputs[0].ctx`.

Comment (sxjscience) [reply python/mxnet/numpy/multiarray.py:323]: Ignore my previous comment. I mean to choose one element in the inputs that has a valid `ctx`.
